### PR TITLE
Modernize cmake and fix a GCC miscompilation bug

### PR DESCRIPTION
**Pull request by jargonzombies**
Fix gcc miscompilation with empty string literal
- All entries in the `symbols` array are empty with gcc 9.3.0 if an empty string literal ("") starts the initializer list. Several tests fail in that case. Adding the zero-termination character fixes this

Modernize CMake, slightly restructure
- Adds a `CMakeLists.txt` to the `external ` subfolder, creating the `catch_main` library that is consumed by the `src/test` subfolder. This allows replacement of `CMAKE_SOURCE_DIR` and `CMAKE_BINARY_DIR` by `CMAKE_CURRENT_...` for more portability.
- Use `target_*` prefixed CMake functions for non-global state mutation (increased portability)
- Reduce list of included directories to enable easier installation of all header sources
  - `config.h` is only needed by the tests, so generate that there and include it there only
  - Change libirc-cross-includes to `libirc`-prefixed includes and remove include/libirc from the includes list. Reduces the list of included directories without regression
  - Move external eigen plugin to within `include/libirc`: From `external/eigen/plugins` to `include/libirc/plugins/eigen`. Removes `external` folder from includes list, allowing cleaner install
- Modern target linking style: Including or linking with absolute paths instead of targets makes installed packages inflexible to relocation of their dependencies
- Add configuration and version files allowing use of `find_package` with installations of libirc

Things to discuss:
- CMake package naming: Maybe you prefer `libirc` to just `irc`?
- Maybe you want the Armadillo/Eigen dependency resolution not to be hardcoded into the installed configuration files. That's possible, too. Then we'd just have to add the dependency resolution to the tests and the `config.cmake.in` file separately.
- CMake package namespacing: Maybe you want the imported package to be namespaced? It's currently not. If someone either includes your project as a subdirectory or via `find_package`, the imported target is always named just `irc`.

### CODE REVIEW

**Review by RMeli — COMMENTED** [review]
Hi @jargonzombies. Thank you very much for your bugfix and the improvements to CMake, they are really appreciated.

---

> All entries in the symbols array are empty with gcc 9.3.0 if an empty string literal ("") starts the initializer list.

Is this behavior documented anywhere? I'll add more compilers to the CI in order to catch such issues early on.

> CMake package naming: Maybe you prefer libirc to just irc?

Either will do. As mentioned in #15, I've been thinking to change the name to the library to RIC (which seems more in line with the literature) or to something more general in case other transformation were added (see #40), but for the time being both libirc and irc would be fine.

> Maybe you want the Armadillo/Eigen dependency resolution not to be hardcoded into the installed configuration files.

Which advantages and disadvantages would the alternative bring?

> CMake package namespacing: Maybe you want the imported package to be namespaced?

Adding a namespace is the best practice in this case? I will CC @peterbygrave here, since this might impact how the library is interfaced in entos (although I suspect their internal version diverged significantly from this repository, so it might not be a problem at all).

File: include/libirc/periodic_table.h [hunk 1]
```diff
@@ -350,20 +350,13 @@ inline constexpr bool valid_atomic_number(std::size_t an) noexcept {
...
+      return i;
     }
   }
 
-  if (an == 0) {
-    throw std::logic_error("Invalid atomic symbol.");
-  }
-
-  return an;
+  throw std::logic_error("Invalid atomic symbol: ");
```
[the remark this answers is not in the document]
Comment (jargonzombies) [reply include/libirc/periodic_table.h:359]: Good catch: This is a remnant from when I was debugging this. I'll add the passed symbol back into the exception string.

**Comment by jargonzombies** [comment]
> Is this behavior documented anywhere? I'll add more compilers to the CI in order to catch such issues early on.

I have to admit I fixed that bug by throwing stuff at the wall until it was resolved. I have a suspicion an empty string literal might be a bugged case for `constexpr` initialization of a `char[]`. The point is for the string to be binary embedded and a pointer to be placed in the array, right? But an empty string literal, what does the compiler do with that? Place a null pointer? I have no idea what's going on in GCC there. Clang seems to deal with it just fine.

> Which advantages and disadvantages would the alternative bring?

Hard to say what constitutes advantages or disadvantages, it depends on what you think is the likeliest use-case for the library I suppose:

I would say an advantage of hardcoding dependencies selected at build/install time into config files would be that the chosen dependency is definitely there when a downstream project calls `find_package`, and no extra variable would have to be set in CMake prior to the call to `find_package` to select armadillo or eigen.

Downside to it is that "package" distributions would be different depending on the used dependency and unflexible in selection of dependencies.

Arguably, it doesn't make much of a difference. In either case, users will have to choose which dependency they want to use. Only if you plan on distributing as flexible a cmake package as possible, then not hardcoding the dependency would be valuable. If every user is going to build/install this manually, selecting their linalg dependency themselves, then I'd say it doesn't matter.

> Adding a namespace is the best practice in this case? 

I'm not qualified to judge that, I think. My personal opinion would be that given that this library exports only a single target, namespacing is arguably overkill, since namespacing is mainly supposed to prevent name conflicts.

File: src/test/CMakeLists.txt [hunk 2]
```diff
@@ -2,69 +2,33 @@
...
+  ${CMAKE_CURRENT_SOURCE_DIR}/irc_test.cpp
+)
+
+foreach(TEST_SOURCE ${TEST_EXECUTABLE_SOURCES})
+  get_filename_component(TEST_NAME ${TEST_SOURCE} NAME_WE)
+  add_executable(${TEST_NAME} ${TEST_SOURCE})
+  target_link_libraries(${TEST_NAME} irc catch_defined)
+  target_include_directories(${TEST_NAME} PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
+  add_test(NAME ${TEST_NAME} COMMAND ${TEST_NAME})
+endforeach()
```
Comment (peterbygrave) [inline src/test/CMakeLists.txt:34]: Nice! I might use a similar technique in other projects

**Comment by peterbygrave** [comment]
No problem with CMake namespacing. Thanks for reminder, I really need to get those additions pushed.

I learnt two new cmake things: `if(TARGET ..)` and `get_filename_component(TEST_NAME ${TEST_SOURCE} NAME_WE)`, thanks 👍

**Comment by RMeli** [comment]
Thank you for the clarifications @jargonzombies

> Arguably, it doesn't make much of a difference. In either case, users will have to choose which dependency they want to use. Only if you plan on distributing as flexible a cmake package as possible, then not hardcoding the dependency would be valuable.

I think at this stage there is not much difference as you mention and therefore there is no real need to change this. But in the future we might want to come back to this and remove the hard-coded dependencies.

> My personal opinion would be that given that this library exports only a single target, namespacing is arguably overkill, since namespacing is mainly supposed to prevent name conflicts.

I tend to agree, it seems overkill. We might come back to this if we will ever change the name of the library. Thanks @peterbygrave for weighing-in.

**Pull request merged**